### PR TITLE
 Introduce `OS.set_drag_mode()`

### PR DESCRIPTION
**Pull request by toger5**
same as #16512 but rebased.

### CODE REVIEW

**Comment by akien-mga** [comment]
Could you squash the commits together?

Note also that it fails building on OSX: https://travis-ci.org/godotengine/godot/jobs/459525810
And there are style issues: https://travis-ci.org/godotengine/godot/jobs/459525806

**Comment by toger5** [comment]
@akien-mga
Couldn't bild on mac myself yesterday so i let travis do it.
I will look into it and squash.

**Comment by toger5** [comment]
@akien-mga fixed the OS X issue and rebased

**Comment by Nobelix** [comment]
What's the status of this?

**Comment by toger5** [comment]
confilcts :( ... no time right now. might check it this weekend

File: platform/osx/os_osx.mm [hunk 1]
```diff
@@ -591,18 +595,75 @@ static void _mouseDownEvent(NSEvent *event, int index, int mask, bool pressed) {
...
+							 display:YES];
+			break;
+		default:
+			[self mouseMoved:event];
+	}
 }
 
 - (void)mouseUp:(NSEvent *)event {
+	if (OS_OSX::singleton->drag_mode == OS::DRAG_MODE_MOVE && [event clickCount] == 2) {
+		OS_OSX::singleton->set_window_maximized(!OS_OSX::singleton->is_window_maximized());
```
Comment (bruvzg) [inline platform/osx/os_osx.mm]: On macOS double-clicking window title should minimize window not maximize, and there's nothing like this implemented on other platforms.

File: platform/windows/os_windows.cpp [hunk 2]
```diff
@@ -915,7 +915,25 @@ LRESULT OS_Windows::WndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+				case DRAG_MODE_MOVE: return HTCAPTION;
+				case DRAG_MODE_RESIZE_TOP: return HTTOP;
+				case DRAG_MODE_RESIZE_RIGHT: return HTRIGHT;
+				case DRAG_MODE_RESIZE_BOTTOM: return HTBOTTOM;
+				case DRAG_MODE_RESIZE_LEFT: return HTLEFT;
+				case DRAG_MODE_RESIZE_TOPLEFT: return HTTOPLEFT;
+				case DRAG_MODE_RESIZE_TOPRIGHT: return HTTOPRIGHT;
+				case DRAG_MODE_RESIZE_BOTTOMRIGHT: return HTBOTTOMRIGHT;
+				case DRAG_MODE_RESIZE_BOTTOMLEFT: return HTBOTTOMLEFT;
+				case DRAG_MODE_NONE: return HTCLIENT;
```
Comment (bruvzg) [inline platform/windows/os_windows.cpp]: `DRAG_MODE_NONE` / `default` should return `DefWindowProcW(hWnd, uMsg, wParam, lParam);` instead, returning `HTCLIENT` will break resizing of normal (non-borderless) windows.

File: platform/osx/os_osx.mm (hunk 1 above)
Comment (bruvzg) [inline platform/osx/os_osx.mm]: On macOS double clicking should minimize window not maximize and this function is not implemented on any other platforms.

**Review by bruvzg — CHANGES_REQUESTED** [review]
On Windows resize is a bit laggy (but same is true for normal Godot window resize).
On Linux/X11 resize lags massively, leave dirty traces and sometimes blinks during resize (it's much more noticeable with transparent window, but happens win non transparent too).

**Comment by bruvzg** [comment]
And another problem, it's possible to resize borderless window to zero size and render it unusable.
Currently there's no way to prevent it (set minimum size limit or process and reject resize events). Ability to set size limits would be useful for normal windows as well.

**Comment by akien-mga** [comment]
There are quite a few outstanding issues that still need to be addressed.

**Comment by toger5** [comment]
Not sure if i find time for it currently. Maybe someone else need to take care of it.

**Comment by aaronfranke** [comment]
@toger5 Is this still desired? If so, this feature would probably need to be redone from scratch, considering the OS and DisplayServer split.

If not, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable). I see that above you say that you may not have time for this, if you can't fix this up yourself, feel free to close it.

**Comment by toger5** [comment]
I guess we close this one. It is not that propable that I will work on this commit. I think a lot it did was fixing issues with the old system. 
It should be redone and the PR can be used as a resource for the person who looks into this.

So I will close for now.

**Pull request closed without merging**